### PR TITLE
c10::Storage python to cpp converter and typecast

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59735
* **#59734**

Adds typecast logic to allow for c10::Storages to cross the Python/C++ barrier with pyBind.

Differential Revision: [D29075279](https://our.internmc.facebook.com/intern/diff/D29075279)

### CODE REVIEW

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@ezyang Are you the proper reviewer for this piece of the code base? I'm not sure what to do for the cpp -> python pybind logic. I couldn't figure out how to build a `PyObject` out of a `at::Storage` since we're missing the dtype. I threw an error, but lmk if there's a better solution that you see

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by ezyang** [comment]
@kurtamohler added you because this might intersect with your ongoing work at https://github.com/pytorch/pytorch/pull/59671

File: torch/csrc/utils/pybind.h [hunk 1]
```diff
@@ -48,6 +48,32 @@ struct type_caster<at::Tensor> {
...
+    return false;
+  }
+
+  static handle
+  cast(const at::Storage& src, return_value_policy /* policy */, handle /* parent */) {
+    TORCH_CHECK(
+        false,
+        "NotImplementedError: pybind conversion of at::Storages from C++ to python not supported.");
+    // Storages are untyped, see: https://github.com/pytorch/pytorch/issues/47442
+    return handle(torch::createPyObject(src, caffe2::TypeMeta()));
```
Comment (ezyang) [inline torch/csrc/utils/pybind.h:73]: Yeah, this is good for now. Then when @kurtamohler is done we can fix this.

**Review by ezyang — APPROVED** [review]

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075279).

**Pull request closed without merging**